Bugfix math params and v0.9.0

DNA386 opened the pull request:
Improve param handling:
- some quick fixes to help fix certain known failures (eg atan2(x,y) -> a*tan*2(x,y))
- added render option to allow math params to render as math in box names
- support latex in circuit names (must be enclosed in $$)

Misc bugfixes:
- phase boxes clickable again

<img width="136" alt="Screenshot 2024-07-09 at 14 04 10" src="https://github.com/CQCL/pytket-circuit-renderer/assets/37022011/98171368-402c-41d4-8980-a5444f3dd298">

Limitations:
- display is cropped to the gate size - if the maths is too big it won't be fully visible. It is left up to the user to pick appropriate names. 
- param rendering is pending a proper sympy parser implementation, the current work-around can lead to unexpected behaviours.

CalMacCQ reviewed (commented):
Rendered maths looks really nice :)  just a few minor comments

CalMacCQ commented on cypress/fixtures/circuits/params.json:
```diff
@@ -0,0 +1,113 @@
+      "args":[["q",[0]],["q",[1]]]
+    },
+    {"op":{"type":"CircBox","box":{"id":"621ee04f-bc1f-4ae3-bbf5-0db8385e9e01","type":"CircBox","circuit":{
+      "name": "$$U_1\\,\\otimes\\,U_2$$",
```
This box doesn't look quite right

Is this just because $U_1\\,\\otimes\\,U_2$ takes up too much space?

<img width="135" alt="Screenshot 2024-07-12 at 12 21 00" src="https://github.com/user-attachments/assets/572ccdb9-aa8e-4bdb-8242-1aaa42338743">

I understand that ensuring expressions are of suitable length is left up to the user

DNA386 replied:
Hmm that shouldn't happen.. (I checked in firefox and chrome and couldn't reproduce it - what version are you using again?)

CalMacCQ replied:
I'm using Firefox Version 128.0 on my M1 Mac

CalMacCQ replied:
This is fixed now. Thanks.

CalMacCQ commented:
Very minor thing

In the simple nested example the Rx gate inside the nested circbox shows "alpha" as a word rather than a symbol. 

Is this intended?

<img width="233" alt="Screenshot 2024-07-12 at 12 27 46" src="https://github.com/user-attachments/assets/37cdbbf1-f3fd-415e-8703-b1d66aba81d3">
<img width="151" alt="Screenshot 2024-07-12 at 12 27 41" src="https://github.com/user-attachments/assets/7ded648d-74d1-4594-aedf-9daf56d30ba6">

CalMacCQ commented on cypress/fixtures/circuits/params.json:
```diff
@@ -0,0 +1,113 @@
+      "args":[["q",[3]],["q",[2]]]
+    },
+    {"op":{"type":"CircBox","box":{"id":"621ee04f-bc1f-4ae3-bbf5-0db8385e9e01","type":"CircBox","circuit":{
+      "name": "<I'm dodgy",
```
I notice that there are a few subcircuits that have a name starting with `<` 

Does this mean that #65 is fixed now? If so thanks :)

DNA386 commented:
> Very minor thing
> 
> In the simple nested example the Rx gate inside the nested circbox shows "alpha" as a word rather than a symbol.
> 
> Is this intended?
> 
> <img alt="Screenshot 2024-07-12 at 12 27 46" width="233" src="https://private-user-images.githubusercontent.com/93673602/348246955-37cdbbf1-f3fd-415e-8703-b1d66aba81d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3ODcwMDEsIm5iZiI6MTcyMDc4NjcwMSwicGF0aCI6Ii85MzY3MzYwMi8zNDgyNDY5NTUtMzdjZGJiZjEtZjNmZC00MTVlLTg3MDMtYjFkNjZhYmE4MWQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEyVDEyMTgyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY2NTMxOTM0MTZlOTlhMTI2YmIwZGM1OWVlZGJhNjY0NGExMDUxYTE3MDZhODQ2Y2MwZGEyYTkwMGIyN2YyMWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gUhfOiaajGotmmlona-VDjJ3k5fhJTK86zCXHpX0vo0"> <img alt="Screenshot 2024-07-12 at 12 27 41" width="151" src="https://private-user-images.githubusercontent.com/93673602/348246963-7ded648d-74d1-4594-aedf-9daf56d30ba6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3ODcwMDEsIm5iZiI6MTcyMDc4NjcwMSwicGF0aCI6Ii85MzY3MzYwMi8zNDgyNDY5NjMtN2RlZDY0OGQtNzRkMS00NTk0LWFlZGYtOWRhZjU2ZDMwYmE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEyVDEyMTgyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3NDkxMjE5MTg2Zjc1OWJiNTJiNTA4ZGIwZmEzMDJmNWQ0ZjEwYWZiYmUxNWM2ZDIzZWQ1OWYzMmRlNzE1NTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NeBbcbmZsxOLChSnt61rwVjmXo9t4zpEivMOuB3pVGE">

This happens because the parameters are getting cropped (but still displaying as math) so the math interpreter doesn't see the whole string and tries to interpret the cropped string directly. I'm still working on improving the parsing, but I can look into making better/smarter crops using it.

Otherwise can potentially disable the cropped param option by default, but this can lead to very long boxes.

DNA386 commented:
Updated to include the parser (still WIP but should be already better than what we had previously). Also added a fix for the scrollbar issue in the param names, and include better param cropping/one-line display.

CalMacCQ reviewed (approved):
Looks good! Approved.

Couple small things

I noticed you changed the spacing for box names. A `Unitary3qBox` now displays as `Unitary 3q Box`. Was this intentional?

Also I noticed that for parameterised gates the name of the gate e.g. Rz is aligned slightly below the parameter value. I don't mind this as its not super noticeable. Would it be possible to align them better? If unclear let me know


<img width="85" alt="Screenshot 2024-07-16 at 14 00 35" src="https://github.com/user-attachments/assets/bd7f974c-6c2e-42be-98d8-4b20d45c5ecc">

DNA386 commented:
> Looks good! Approved.
> 
> Couple small things
> 
> I noticed you changed the spacing for box names. A `Unitary3qBox` now displays as `Unitary 3q Box`. Was this intentional?
> 
> Also I noticed that for parameterised gates the name of the gate e.g. Rz is aligned slightly below the parameter value. I don't mind this as its not super noticeable. Would it be possible to align them better? If unlcear let me know
> 
> <img alt="Screenshot 2024-07-16 at 14 00 35" width="85" src="https://private-user-images.githubusercontent.com/93673602/349093483-bd7f974c-6c2e-42be-98d8-4b20d45c5ecc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExMzU1MzUsIm5iZiI6MTcyMTEzNTIzNSwicGF0aCI6Ii85MzY3MzYwMi8zNDkwOTM0ODMtYmQ3Zjk3NGMtNmMyZS00MmJlLTk4ZDgtNGIyMGQ0NWM1ZWNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE2VDEzMDcxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiYzFlODg5ZjI2ZTk0YTE4M2QyOTRkNWQ5ZjdhMGVmMWY5ZDNhYjhkNDYxZjdlMDVkY2E4NGU4ODZkODMzZGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XRxOUJVRjRSovEoKl4C9EFzbTebIcrAqps2dCE2jFJM">

The change to 'Unitary 3q Box' was an attempt to make the names nicer to read, but I can change it back if you don't think it would be helpful.

DNA386 commented:
For the alignment thing, I think its to do with how mathjax sets the baseline for the text (which annoyingly happens to be above the non-math text) - there's not really anything I can do to fix it other than forcing it to look like math, but that then interferes with the box names that have math in them

CalMacCQ commented:
> The change to 'Unitary 3q Box' was an attempt to make the names nicer to read, but I can change it back if you don't think it would be helpful.


I have a slight prefernce for the name `Unitary3qBox` vs `Unitary 3q Box` as its the name of the python class. I'll leave it up to you though, not a big deal.
